src/pyclass.rs

```diff
@@ -0,0 +1,642 @@
+            return;
+        }
+
+        match Self::type_object().tp_free {
+            Some(free) => free(obj as *mut c_void),
+            None => tp_free_fallback(obj),
+        }
+    }
+}
+
+#[doc(hidden)]
+pub unsafe fn tp_free_fallback(obj: *mut ffi::PyObject) {
+    let ty = ffi::Py_TYPE(obj);
+    if ffi::PyType_IS_GC(ty) != 0 {
+        ffi::PyObject_GC_Del(obj as *mut c_void);
+    } else {
+        ffi::PyObject_Free(obj as *mut c_void);
+    }
+
+    // For heap types, PyType_GenericAlloc calls INCREF on the type objects,
+    // so we need to call DECREF here:
+    if ffi::PyType_HasFeature(ty, ffi::Py_TPFLAGS_HEAPTYPE) != 0 {
+        ffi::Py_DECREF(ty as *mut ffi::PyObject);
+    }
+}
+
+/// If `PyClass` is implemented for `T`, then we can use `T` in the Python world,
+/// via `PyClassShell`.
+///
+/// `#[pyclass]` attribute automatically implement this trait for your Rust struct,
+/// so you don't have to use this trait directly.
+pub trait PyClass:
+    PyTypeInfo<ConcreteLayout = PyClassShell<Self>> + Sized + PyClassAlloc + PyMethodsProtocol
+{
+    type Dict: PyClassDict;
+    type WeakRef: PyClassWeakRef;
+}
+
+unsafe impl<T> PyTypeObject for T
+where
+    T: PyClass,
+{
+    fn init_type() -> NonNull<ffi::PyTypeObject> {
+        let type_object = unsafe { <Self as PyTypeInfo>::type_object() };
+
+        if (type_object.tp_flags & ffi::Py_TPFLAGS_READY) == 0 {
+            // automatically initialize the class on-demand
+            let gil = Python::acquire_gil();
+            let py = gil.python();
+
+            initialize_type::<Self>(py, <Self as PyTypeInfo>::MODULE).unwrap_or_else(|e| {
+                e.print(py);
+                panic!("An error occurred while initializing class {}", Self::NAME)
+            });
+        }
+
+        unsafe { NonNull::new_unchecked(type_object) }
+    }
+}
+
+/// `PyClassShell` represents the concrete layout of `PyClass` in the Python heap.
+///
+/// You can use it for testing your `#[pyclass]` correctly works.
+///
+/// ```
+/// # use pyo3::prelude::*;
+/// # use pyo3::{py_run, PyClassShell};
+/// #[pyclass]
+/// struct Book {
+///     #[pyo3(get)]
+///     name: &'static str,
+///     author: &'static str,
+/// }
+/// let gil = Python::acquire_gil();
+/// let py = gil.python();
+/// let book = Book {
+///     name: "The Man in the High Castle",
+///     author: "Philip Kindred Dick",
+/// };
+/// let book_shell = PyClassShell::new_ref(py, book).unwrap();
+/// py_run!(py, book_shell, "assert book_shell.name[-6:] == 'Castle'");
+/// ```
+#[repr(C)]
+pub struct PyClassShell<T: PyClass> {
+    ob_base: <T::BaseType as PyTypeInfo>::ConcreteLayout,
+    pyclass: ManuallyDrop<T>,
+    dict: T::Dict,
+    weakref: T::WeakRef,
+}
+
+impl<T: PyClass> PyClassShell<T> {
+    /// Make new `PyClassShell` on the Python heap and returns the reference of it.
+    pub fn new_ref(py: Python, value: impl IntoInitializer<T>) -> PyResult<&Self>
```

Comment:
I'm trying to get my head around the refcounting here.

The return type is `&PyClassShell<T>`, so presumably this is a borrowed reference? Is the refcount of the allocated python object 1 or 0 at this point?

Reply:
We have **really** complicated system.
See #679 for detail.